📖 update clusterctl provider contract for clusterclass support (#5582)
kubernetes-sigs/cluster-api: docs/book/src/clusterctl/provider-contract.md

Additionally, the value of the command argument to `clusterctl generate cluster <cluster-name>` (`<cluster-name>` in this case), will
be applied to every occurrence of the `${ CLUSTER_NAME }` variable.

### ClusterClass definitions

> An infrastructure provider could publish a **cluster class definition** file to be used by `clusterctl generate cluster` that will be used along

Reviewer: nit: let's use consistently ClusterClass given that it is an API type
```suggestion
An infrastructure provider could publish a **ClusterClass definition** file to be used by `clusterctl generate cluster` that will be used along
```

with the workload cluster templates.

> This is a single YAML with _all_ the objects required that make up the cluster class.

Reviewer 2: Should we consistently use `ClusterClass` ?

The following rules apply:

#### Naming conventions

ClusterClass definitions MUST be stored in the same folder as the component YAML and follow this naming convention:

> 1. The cluster class definition should be names `clusterclass-{class-name}.yaml`, e.g `clusterclass-prod.yaml`

Reviewer 2: ```suggestion
1. The cluster class definition should be named `clusterclass-{ClusterClass-name}.yaml`, e.g `clusterclass-prod.yaml`
```
I think you can inline l.261 by slightly changing the var name

`{class-name}` is the name of ClusterClass. 

> Each provider should create user facing documentation with the list of available cluster class definitions.

Author: I have added an additional note along these lines under the Variables section. Seemed more appropriate to add it under that section.

#### Target namespace

> The cluster class definition YAML MUST asume the target namespace already exists.

Reviewer 3: ```suggestion
The cluster class definition YAML MUST assume the target namespace already exists.
```

> All objects in the cluster class definition YAML MUST be deploymned in the same namespace.

Reviewer 2: ```suggestion
All objects in the cluster class definition YAML MUST be deployed in the same namespace.
```
I wonder if we should recommend that the namespace is not set.

What would happen if someone sets a specific namespace? Would it be overwritten automatically or only when `${NAMESPACE}` is used?

Reviewer 2: Maybe, yup. But not if it is set in the refs in the ClusterClass.

Reviewer 2: Given that we don't want to recommend using variables in ClusterClass we should probably recommend that the namespace should not be set at all.

So that then the combination of fixTargetNamespace and our defaulting webhook will make sure the right namespace is used. Otherwise things can get messy if the top-level namespace is overwritten and is not the same as the one hard-coded in the refs.

Author: >What would happen if someone sets a specific namespace? Would it be overwritten automatically or only when ${NAMESPACE} is used?

It will be overridden automatically. 

As long as we have a note saying that none of the references in the ClusterClass set the namespace things should work as expected. (Will add this note in the contract.)

>Given that we don't want to recommend using variables in ClusterClass we should probably recommend that the namespace should not be set at all.

We can definitely add this recommendation to make things consistent. But just FYI, it is totally fine if a namespace is specified in the resource. As long as none of the references use a namespace it will work.

Reviewer 2: Got it, thx

> It is strongly recommended that none of the objects in the cluster class definitions are shared acorss multiple definitions.

Reviewer 2: ```suggestion
It is strongly recommended that none of the objects in the cluster class definitions are shared across multiple definitions.
```
nit

Reviewer: ```suggestion
Even if technically possible, it is strongly recommended that none of the objects in the ClusterClass definitions are shared acorss multiple definitions; this helps on preventing that changing on an object impacts inadvertently many ClusterClasses, and consequently, all the Clusters using those.
```

Reviewer 4: If this something we want to prevent from happening and validate with webhooks?

Reviewer: > If this something we want to prevent from happening and validate with webhooks?

Advanced users can choose to opt-in and reusing templates, it could be a powerful feature in the proper hands.
However, but given that in this context are talking about template for the quick start UX let reccomand the safest path.

Author: >If this something we want to prevent from happening and validate with webhooks?

We would not want to prevent this as a feature in ClusterClass. As @fabriziopandini mentioned reusing templates can become a really powerful feature that advanced users will want to use. 

Given that this doc mostly focuses on 'example' ClusterClass definitions that a provider can provide, which in most cases will be used for quick start, we would want to recommend the safest path.

> #### Variables

Reviewer: This contrast with L263 ...

Author: I dont see how this is in contrast with L263. L263 just says that the providers should document the list of available ClusterClass definitions.

> Currently the cluster class definitions SHOULD NOT have any variables in it.

Reviewer 3: ```suggestion
Currently the cluster class definition SHOULD NOT have any variables in it.
```
or 
```suggestion
Currently the cluster class definitions SHOULD NOT have any variables in them.
```

#### Note

A cluster class definition is automatically included (if not already installed in the cluster) in the output of  `clusterctl generate cluster` 

> if the workload cluster uses a managed topology.

Reviewer 2: ```suggestion
if the cluster template uses a managed topology.
```
nit, not sure what fits better

Reviewer: ```suggestion
if the workload cluster uses a managed topology and a ClusterClass with the same name does not already exists in the Cluster
```

1. The ClusterClass definition should be named `clusterclass-{ClusterClass-name}.yaml`, e.g `clusterclass-prod.yaml`.

`{ClusterClass-name}` is the name of the ClusterClass that is referenced from the Cluster.spec.topology.class field 
in the Cluster template; Cluster template files using a ClusterClass are usually simpler because they are no longer 
required to have all the templates.

Each provider should create user facing documentation with the list of available ClusterClass definitions.

#### Target namespace

The ClusterClass definition YAML MUST assume the target namespace already exists.

The references in the ClusterClass definition should NOT specify a namespace.

It is recommended that none of the objects in the ClusterClass YAML should specify a namespace.

Even if technically possible, it is strongly recommended that none of the objects in the ClusterClass definitions are shared acorss multiple definitions; 
this helps in preventing changing an object inadvertently impacting many ClusterClasses, and consequently, all the Clusters using those ClusterClasses.

#### Variables

> Currently the ClusterClass definitions SHOULD NOT have any variables in them.
> 
> ClusterClass definitions files should not use variable substitution, given that ClusterClass and managed topologies provide an alternative model for variable definion. 

Reviewer 2: ```suggestion
Currently the ClusterClass definitions SHOULD NOT have any variables in them.

ClusterClass definitions files should not use variable substitution, given that ClusterClass and managed topologies provide an alternative model for variable definition. 
```

> ClusterClass definitions files should not use variable substitution, given that ClusterClass and managed topologies provide an alternative model for variable definion. 

Author: I read it as the first sentence mentioning that ClusterClasses should not have variables (envsubst style variables as used in workload cluster templates) in them and the second sentence explains why we should not use variables in ClusterClass definitions. 

>Not sure if it is intended, but this sounds like the ClusterClasses should also not use variables (the new ones) as it writes that variables and variable substitutions should not be used.

By "the new ones" I am assuming you mean the variables used by for patching. If so, the doc is not talking about them.
How about "Currently the ClusterClass definitions SHOULD NOT have any environment variables in them." to avoid ambiguity.?

> ClusterClass definitions files should not use variable substitution, given that ClusterClass and managed topologies provide an alternative model for variable definion. 

Reviewer 2: Sounds great! My problem was only that the there's no way to tell with the current wording if the variables in the first sentence are ClusterClass variables.
